packages/lib/gpcircuits/circuits/proto-pod-gpc.circom

```diff
@@ -108,9 +109,61 @@ template ProtoPODGPC (
         entryValueHashes[i] <== entryProofSiblings[i][0];
     }
 
-    // Entry can be compared for equality (by hash) to another entry (by index).
-    // This can be disabled by comparing to self: entryEqualToOtherEntryIndex[i] = i
-    signal input entryEqualToOtherEntryByIndex[MAX_ENTRIES];
+    /*
+     * Virtual entries. These are entries derived from PODs'
+     * cryptographic data, e.g. the signers' public keys.
+     *
+     * This block pipes them in and forms signals to be fed into the
+     * entry constraint, tuple and list membership modules below, as well
+     * as those revealing their hashes where applicable.
+     */
+    
+    // Maximum number of virtual entries.
+    var MAX_VIRTUAL_ENTRIES = MAX_OBJECTS;
+
+    // Virtual entry value hashes deduced from cryptographic data. At
+    // the moment, this consists of hashed object signers' public
+    // keys.
+    signal virtualEntryValueHashes[MAX_VIRTUAL_ENTRIES];
+        for (var i = 0; i < MAX_OBJECTS; i++) {
+        virtualEntryValueHashes[i]
+            <== Poseidon(2)([objectSignerPubkeyAx[i], objectSignerPubkeyAy[i]]);
+    }
+    
+    // Boolean flags for virtual entry behaviour.
+    signal input virtualEntryIsValueHashRevealed /*MAX_VIRTUAL_ENTRIES packed bits*/;
+    signal virtualEntryIsValueHashRevealedBits[MAX_VIRTUAL_ENTRIES]
+        <== Num2Bits(MAX_VIRTUAL_ENTRIES)(virtualEntryIsValueHashRevealed);
+
+    // Virtual entry value is optionally revealed, or set to -1 if
+    // not.
+    signal output virtualEntryRevealedValueHash[MAX_VIRTUAL_ENTRIES];
+    for (var i = 0; i < MAX_VIRTUAL_ENTRIES; i++) {
+        virtualEntryRevealedValueHash[i]
+            <== ValueOrNegativeOne()(
+                virtualEntryValueHashes[i],
+                virtualEntryIsValueHashRevealedBits[i]);
+    }
+
+    // Append virtual entry hashes to entry hashes for use in the
+    // entry constraint module.
+    signal entryAndVirtualEntryValueHashes[MAX_ENTRIES + MAX_VIRTUAL_ENTRIES]
+        <== Append(MAX_ENTRIES, MAX_VIRTUAL_ENTRIES)(
+            entryValueHashes,
+            virtualEntryValueHashes);
+
+    /*
+     * 1 EntryConstraintModule for each entry, virtual or otherwise.
+     */
+    
+    // Entry or virtual entry can be compared for equality (by hash)
+    // to another entry or virtual entry (by index). An index less
+    // than MAX_ENTRIES refers to an entry and an index i in
+    // [MAX_ENTRIES, MAX_ENTRIES + MAX_VIRTUAL_ENTRIES[ refers to the
+    // (i - MAX_ENTRIES)th virtual entry.
+    // This can be disabled by comparing to self:
+    //   entryEqualToOtherEntryIndex[i] = i
+    signal input entryEqualToOtherEntryByIndex[MAX_ENTRIES + MAX_VIRTUAL_ENTRIES];
```

Comment:
This sum expression (MAX_ENTRIES + MAX_VIRTUAL_ENTRIES) appears so many times below it deserves a name, not just for the variable but the concept.  Is it just TOTAL_ENTRIES or ALL_ENTRIES?  Or is there a name for an entry which is either real or virtual?  It would be clearer than names like entryAndVirtualEntry... below.

Reply:
Agreed.